File: PS5/HW5.R
############## Stat 243 ps5 ###################
# Meng Wang, 21706745
# October 18
setwd("./Stat243/Homework/PS5")
rm(list = ls())
library(pryr)
############# Problem 1(a) ###################
# check the base and smallest float it can reach
.Machine$double.base
.Machine$double.ulp.digits

# check whehter 1e-16 is reachable or not
num <- 1.000000000001
options(digits = 16)
num
############# Problem 1(b) ###################
x <- c(1, rep(1e-16, times = 10000))
options(digits = 20)
sum(x)

############# Problem 1(d) ###################
# use for loop, first 1 is at the first beginning
d1_sum <- 0.0
for(i in 1:length(x)){
  d1_sum = d1_sum + x[i]
}
options(digits = 20)
d1_sum

# now add 1 at the end
d2_sum <- 0.0
for(i in 1:10000){
  d2_sum = d2_sum + 1e-16
}
d2_sum = d2_sum + 1.0
options(digits = 20)
d2_sum


############# Problem 2 ###################

int_A <- matrix(as.integer(13), nrow = 1e4, ncol = 1e4)
double_A <- matrix(as.numeric(13.0), nrow = 1e4, ncol = 1e4)
# matrix transform
system.time( t(int_A))
system.time(t(double_A))
# find the max value
system.time( max(int_A ))
system.time( max(double_A) )

# creat the pdf file
library(knitr); 
knit2pdf('HW5.Rtex')

File: PS5/HW5.py
# Stat243 PS5
# Meng Wang, SID:21706745
# Oct 18, 2015

import numpy as np

####### Problem 1(c) ###########
vec = np.array([1e-16]*(10001))
vec[0] = 1.0
print "========== 1c ============="
print '%.20f' % np.sum(vec)

####### Problem 1(d) ###########
# use for loop, but add 1 at the beginning
d1_sum = 0.0
d1_sum = d1_sum + 1.0
for i in range(10000):
	d1_sum = d1_sum + 1e-16
print "========== 1d ============="
print "====== Add 1 first ====="
print '%.20f' % d1_sum

# use for loop, add 1 at the end
d2_sum = 0.0
for i in range(10000):
        d2_sum = d2_sum + 1e-16
d2_sum = d2_sum + 1.0
print "====== Add 1 last ====="
print '%.20f' % d2_sum



